add New Emu Checklist (#28)
RetroAchievements/RAEmus: NewEmuChecklist.txt

RetroAchievements Emulator Requirements:
* If the emulator does not support a feature, the requirement is automatically met. This is primarily a set of
  guidelines to ensure the interaction between the emulator and the toolkit is consistent between emulators.
* If the system is supported by RetroArch, explain why a standalone emulator is desirable. RetroArch support is 
  preferred as it is usable by more players (i.e. not just Windows users), and adding support to RALibRetro for 
  an additional core is much easier than overhauling an entire emulator.

[ ] Game Identification
  [ ] Hash method documented and approved
  [ ] If emulator supports dynamic patching [i.e. ips], hash is calculated after applying patch

>   [ ] Hash ignores any persistant save data [SRAM]

Reviewer: ```suggestion
  [ ] Hash ignores any persistent save data [SRAM]
```

>   [ ] Game can be properly identified (RA_OnLoadNewRom)

Reviewer: ```suggestion
  [ ] Game can be properly identified (RA_OnLoadNewRom and/or RA_IdentifyRom and RA_ActivateGame)
```

>   [ ] Unidentified game provides some descriptor [typically filename] to unknown game dialog (RA_InstallSharedFunctions>RA_GetEstimatedGameTitle)

Reviewer: For reference, I don't think this is done for many current emulators.

Author: Acknowledged. I plan to run through the checklist on all of the existing emulators over the next few weeks.

>   [ ] Unidentified game provides list of known games for console (RA_SetConsoleID)

Reviewer: That is typically handled by `RAIntegration`, unless the emulator supports multiple platforms; an indication of that would be helpful.

[ ] Emulator

>   [ ] DLL is downloaded if not available (RA_Init - new enum may be needed)

Reviewer: ```suggestion
  [ ] DLL is downloaded if not available (RA_Init - new EmulatorID enum value may be needed in RA_Interface.h)
```

>   [ ] RetroAchievements menu is visible (RA_CreatePopupMenu) and functional (RA_InvokeDialog)

Reviewer: I believe that this is all handled by `RAIntegration`.

Author: nope, RA_CreatePopupMenu returns a handle that must be added to the application's menu bar: https://github.com/RetroAchievements/RAEmus/blob/master/RAQUASI88/src/WIN32/retroachievements.cpp#L141

And RA_InvokeDialog must be called from the applications WndProc: https://github.com/RetroAchievements/RAEmus/blob/master/RAQUASI88/src/WIN32/retroachievements.cpp#L334

Reviewer: > RA_CreatePopupMenu returns a handle that must be added to the application's menu bar

I see, must have missed it while copying and pasting this function. :)

  [ ] RetroAchievements menu updates when logging in/out (RA_InstallSharedFunctions>RA_RebuildMenu)
  [ ] Login dialog displayed on first run or auto-login on future runs (RA_AttemptLogin)
  [ ] Title includes logged in user (RA_UpdateAppTitle)

>   [ ] RALog.txt ends with "Shutdown complete" after closing emulator (RA_Shutdown)

Reviewer: Just nitpicking, but I think behavior (e.g. RA resources are disposed on exit) would act as better criteria than logging output.

Author: disposing the resources is definitely the intended reason for calling `RA_Shutdown`, but the log message is the easiest way to verify that it's happening.

  [ ] Full-window overlay can be displayed (RA_SetPaused)
  [ ] Full-window overlay can be navigated (RA_UpdateRenderOverlay>pInput)
  [ ] Full-window overlay navigation uses player's custom controller bindings, not hard-coded inputs.
  [ ] Closing the full-window overlay unpauses the emulator (RA_InstallSharedFunctions>RA_CauseUnpause)
  [ ] Resetting the emulator disables any active popups/leaderboards (RA_OnReset)

>   [ ] Performance of emulator (without any debugging windows open) should not be noticably different than the source emulator

Reviewer: ```suggestion
  [ ] Performance of emulator (without any debugging windows open) should not be noticeably different than the source emulator at real-time speed
```

Author: will use "100% speed" to be consistent with other items

>   [ ] Emulator must perform at intended framerate on an "average" machine.

Reviewer: Defining "average" might be helpful.

Author: It would, but I don't have that metric. It's something GameDragon requested I add. Apparently some of the attempted previous emulators ran like crap even before trying to integration RA.

Author: I've decided to say "a PC that could have been purchased at a retail store within the last four years". It's still pretty subjective as there's a big difference between what Walmart sells and what Best Buy sells.

[ ] Achievements/Leaderboards/Rich Presence
  [ ] Achievements can be created (RA_InstallSharedFunctions>RA_GameIsActive)
  [ ] Achievements trigger (RA_InstallMemoryBank>pReader, RA_DoAchievementsFrame, RA_HandleHTTPResults)
  [ ] Achievement trigger displays notification (RA_UpdateRenderOverlay)
  [ ] Achievement can pause emulator [Pause on Reset, Pause on Trigger] (RA_InstallSharedFunctions>RA_CausePause)

>   [ ] Achievements are not processed when emulator is paused [HitCount for true condition does not increment]

Reviewer: If based on `GameIsActive()`, please specify. Also, for reference, this is not at all implemented on some (most?) emulators currently.

Author: It's based on RA_DoAchievementsFrame. I'll update the comment.

  [ ] When using frame advance, achievements are only processed when the frame is advanced.
  [ ] Changing the current game prompts the user to save modifications (RA_ConfirmLoadNewRom)
    [ ] Choosing not to save modifications aborts changing the current game.
  [ ] Closing the emulator prompts the user to save modifications (RA_ConfirmLoadNewRom)
    [ ] Choosing not to save modifications aborts closing the emulator.
  [ ] Closing the current game prompts the user to save modifications (RA_ConfirmLoadNewRom)
    [ ] Choosing not to save modifications aborts closing the current game.
  [ ] Closing the current game clears the list of achievements for the game (RA_OnLoadNewRom(NULL,0))

> [ ] Softcore

Reviewer: In general, functionality that is allowed in softcore should not need to be explicitly specified in my opinion, as long as the `Hardcore` section is clear.

Author: You'd think so, but if the hardcore check is not implemented correctly, it causes issues in softcore. Having the list just makes sure the functionality is tested.

  [ ] Save states can be created

>   [ ] Save States can be loaded

Reviewer: ```suggestion
  [ ] Save states can be loaded
```
